src/streaming/downsampler.cpp

```diff
@@ -0,0 +1,295 @@
+#include "downsampler.hh"
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "zarrv3.array.writer.hh"
+
+namespace {
+template<typename T>
+[[nodiscard]] ByteVector
+scale_image(ConstByteSpan src, size_t& width, size_t& height)
+{
+    const auto bytes_of_src = src.size();
+    const auto bytes_of_frame = width * height * sizeof(T);
+
+    EXPECT(bytes_of_src >= bytes_of_frame,
+           "Expecting at least ",
+           bytes_of_frame,
+           " bytes, got ",
+           bytes_of_src);
+
+    const int downscale = 2;
+    constexpr auto bytes_of_type = static_cast<double>(sizeof(T));
+    const double factor = 0.25;
+
+    const auto w_pad = static_cast<double>(width + (width % downscale));
```

Comment:
Use of `auto` is not making this code clearer. The many uses of static_cast and reinterpret_cast also look like something is wrong. Cast to what you want at the top of the function/outside the loops so things are already in the right types.